### PR TITLE
Use block cache deleters for type inspection and checking

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Along with #8225, this is working toward gathering statistics
about the kinds of items in block cache by iterating over the cache
during periodic stats dump. We need some way to flag what kind of
blocks each entry belongs to, preferably without changing the Cache API.
One of the complications is that Cache is a general interface that could
have other users that don't adhere to whichever convention we decide
on for keys and values. Or we would pay for an extra field in the Handle
that would only be used for this purpose.

This change uses a back-door approach, the deleter, to mark the kind
of Cache entries. This has the added benefit of ensuring that each entry
we check the kind of has the proper code origin; if the entry came from
some other part of the code, it will use an unrecognized deleter, which
we can simply attribute to an "unknown" bucket.

In detail, this change pulls the awkward BlocklikeTraits into
CachableEntry, along with an API for translating between a deleter and
`<CacheEntryType, BlockType>`. This should soon be used for collecting
stats.

A side benefit/feature in this change is that the information in the
deleter can be used for "type checking" block cache accesses, through
a new option extra_block_cache_checks. Any logical error that could lead
to a block cache key collision can break type safety, such as with a
bad cast from Block to BlockContents. This option detects and reports
such errors before the bad cast. This is now always enabled for
db_stress, and for one CircleCI unit test configuration, to ensure we
stay clean, but otherwise disabled by default because of modest
performance penalty.

Test Plan: TODO: small unit tests and performance tests

### CODE REVIEW

File: table/block_based/cachable_entry.cc [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+// Using a named local namespace to (hopefully) help ensure that different
+// instantiations of UniqueDeleter have unique fnptr addresses.
+namespace cachable_entry_details {
+
+template <class T, BlockType kBlockType>
+void UniqueDeleter(const Slice& /*key*/, void* value) {
+  auto entry = reinterpret_cast<T*>(value);
+  delete entry;
+}
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/block_based/cachable_entry.cc:29]: That's not the plan, instead "gathering statistics about the kinds of items in block cache by iterating over the cache during periodic stats dump," referring to `InternalStats`

File: table/block_based/block_based_table_reader.cc [hunk 2]
```diff
@@ -73,72 +74,6 @@ BlockBasedTable::~BlockBasedTable() {
 
 std::atomic<uint64_t> BlockBasedTable::next_cache_key_id_(0);
 
-template <typename TBlocklike>
-class BlocklikeTraits;
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:77]: I think I agree with your suggestion on noting that CachableEntry does not have any specific relationship to block-based table (despite being in table/block-based and only being instantiated for it) and noting that my extensions are essentially all static and specific to block-based table. But I still intend to reduce LoC in block_based_table_reader.cc ;)

[the remark this answers is not in the document]
Comment (ltamasi) [reply table/block_based/block_based_table_reader.cc:77]: Sure, `block_based_table_reader.cc` is still definitely too crowded. How about keeping the classes separate and just moving `BlocklikeTraits` to its own file(s)?

File: table/block_based/cachable_entry.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ltamasi) [reply table/block_based/cachable_entry.cc:29]: Got it, so you're just using the deleters to encode type information

**Comment by pdillinger** [comment]
Replaced by #8297

**Pull request closed without merging**